domain/modelagent/service/service.go

```diff
@@ -24,43 +52,247 @@ type Service struct {
+		return version.Zero, errors.Errorf(
+			"machine %q does not exist", machineName,
+		).Add(machineerrors.MachineNotFound)
+	} else if err != nil {
+		return version.Zero, errors.Errorf(
+			"checking if machine %q exists when getting target agent version: %w",
+			machineName, err,
+		)
 	}
+
+	modelUUID, err := s.st.GetModelUUID(ctx)
+	if err != nil {
+		return version.Zero, errors.Errorf(
+			"getting machine %q model uuid: %w", machineName, err,
+		)
+	}
+
+	return s.Service.st.GetModelTargetAgentVersion(ctx, modelUUID)
 }
 
-// GetModelAgentVersion returns the agent version for the current model.
+// GetModelTargetAgentVersion returns the agent version for the specified model.
 // The following errors can be returned:
 //   - [errors.NotValid] if the model ID is not valid;
 //   - [github.com/juju/juju/domain/model/errors.NotFound] if no model exists
 //     for the provided ID.
-func (s *ModelService) GetModelAgentVersion(ctx context.Context) (version.Number, error) {
-	return s.Service.GetModelAgentVersion(ctx, s.modelID)
+func (s *Service) GetModelTargetAgentVersion(
+	ctx context.Context,
+	modelID model.UUID,
+) (version.Number, error) {
+	if err := modelID.Validate(); err != nil {
+		return version.Zero, errors.Errorf("validating model ID: %w", err)
+	}
+	return s.st.GetModelTargetAgentVersion(ctx, modelID)
+}
+
+// GetModelTargetAgentVersion returns the target agent version for the entire
+// model. The following errors can be returned:
+// - [github.com/juju/juju/domain/model/errors.NotFound] - When the model does
+// not exist.
+func (s *ModelService) GetModelTargetAgentVersion(
+	ctx context.Context,
+) (version.Number, error) {
+	modelUUID, err := s.st.GetModelUUID(ctx)
+	if err != nil {
+		return version.Zero, errors.Errorf("getting the current model's uuid: %w", err)
+	}
+
+	return s.Service.GetModelTargetAgentVersion(ctx, modelUUID)
+}
+
+// GetUnitTargetAgentVersion reports the target agent version that should be
+// being run on the provided unit identified by name. The following errors
+// are possible:
+// - [applicationerrors.UnitNotFound] - When the unit in question does not exist.
+// - [github.com/juju/juju/domain/model/errors.NotFound] - When the model the
+// unit belongs to no longer exists.
+func (s *ModelService) GetUnitTargetAgentVersion(
+	ctx context.Context,
+	unitName string,
+) (version.Number, error) {
+	err := s.st.CheckUnitExists(ctx, unitName)
+	if errors.Is(err, applicationerrors.UnitNotFound) {
+		return version.Zero, errors.Errorf(
+			"unit %q does not exist", unitName,
+		).Add(applicationerrors.UnitNotFound)
+	} else if err != nil {
+		return version.Zero, errors.Errorf(
+			"checking if unit %q exists when getting target agent version: %w",
+			unitName, err,
+		)
+	}
+
+	modelUUID, err := s.st.GetModelUUID(ctx)
+	if err != nil {
+		return version.Zero, errors.Errorf(
+			"getting unit %q model uuid: %w", unitName, err,
+		)
+	}
+
+	return s.Service.st.GetModelTargetAgentVersion(ctx, modelUUID)
```

Comment:
I'm not really a fan of dipping into another types internal state. I wonder if it would have been better to lift that state into the `ModelService` directly. That way if the underlying composed Service changes, it doesn't break this Service.

Reply:
As discussed we are swapping DB's for model agent so this will all correct it self very shortly.